rpcserver.go

```diff
@@ -1903,6 +1904,89 @@ func handleGetBlock(s *rpcServer, cmd interface{}, closeChan <-chan struct{}) (i
 	return blockReply, nil
 }
 
+// handleGetBlockchainInfo implements the getblockchaininfo command.
+func handleGetBlockchainInfo(s *rpcServer, cmd interface{}, closeChan <-chan struct{}) (interface{}, error) {
+	params := s.server.chainParams
+	best := s.chain.BestSnapshot()
+	dInfo := make(map[string]dcrjson.AgendaInfo)
+
+	// Fetch the current chain work using the the best block hash.
+	chainWork, err := s.chain.ChainWork(&best.Hash)
+	if err != nil {
+		return nil, rpcInternalError(err.Error(), "Could not fetch chain work.")
+	}
+
+	// Guess the syncing status of the node.
+	initialBlockDownload := !s.chain.IsCurrent()
+
+	// Estimate the verification progress of the node.
+	syncHeight := s.server.blockManager.SyncHeight()
+	var verifyProgress float64
+	switch syncHeight > 0 {
+	case false:
+		verifyProgress = 0
+	case true:
+		verifyProgress = float64(best.Height) / float64(syncHeight)
+	}
+
+	// Fetch the maximum allowed block size.
+	maxBlockSize, err := s.chain.MaxBlockSize()
+	if err != nil {
+		return nil, rpcInternalError(err.Error(),
+			"Could not fetch max block size.")
+	}
+
+	// Fetch the agendas of the consensus deployments as well as their
+	// threshold states and state activation heights.
+	for version, deployments := range params.Deployments {
+		for _, agenda := range deployments {
+			aInfo := dcrjson.AgendaInfo{
+				StartTime:  agenda.StartTime,
+				ExpireTime: agenda.ExpireTime,
+			}
+
+			state, err := s.chain.NextThresholdState(&best.PrevHash, version,
+				agenda.Vote.Id)
+			if err != nil {
+				return nil, rpcInternalError(err.Error(),
+					fmt.Sprintf("Could not fetch the next threshold state "+
+						"for agenda with id (%v).", agenda.Vote.Id))
+			}
+
+			if state.State != blockchain.ThresholdDefined {
```

Comment:
No reason to get the state and avoid calling `StateLastChangedHeight` if the state is defined.  The function will just return 0 in that case, which is the same result.